tests/Unit/DataStructures/Tensor/Expressions/Test_AddSubSymmetry.cpp

```diff
@@ -0,0 +1,202 @@
+
+#include "Framework/TestingFramework.hpp"
+
+#include <cstddef>
+#include <type_traits>
+
+#include "DataStructures/Tensor/Symmetry.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace {
+template <auto&... TensorIndices>
+using make_tensorindex_list =
+    tmpl::list<std::decay_t<decltype(TensorIndices)>...>;
+
+void test_rank0() {
+  using symm = Symmetry<>;
+  using tensorindex_list = make_tensorindex_list<>;
+
+  CHECK(
+      std::is_same_v<typename TensorExpressions::detail::AddSubSymmetry<
+                         symm, symm, tensorindex_list, tensorindex_list>::type,
+                     tmpl::integral_list<std::int32_t>>);
+}
+
+void test_rank1() {
+  using symm = Symmetry<1>;
+  using tensorindex_list = make_tensorindex_list<ti_a>;
+
+  CHECK(
+      std::is_same_v<typename TensorExpressions::detail::AddSubSymmetry<
+                         symm, symm, tensorindex_list, tensorindex_list>::type,
+                     tmpl::integral_list<std::int32_t, 1>>);
+}
+
+void test_rank2() {
+  using symmetric_symm = Symmetry<1, 1>;
+  using asymmetric_symm = Symmetry<2, 1>;
+  using tensorindex_list_ij = make_tensorindex_list<ti_i, ti_j>;
+  using tensorindex_list_ji = make_tensorindex_list<ti_j, ti_i>;
+
+  CHECK(std::is_same_v<typename TensorExpressions::detail::AddSubSymmetry<
+                           symmetric_symm, symmetric_symm, tensorindex_list_ij,
+                           tensorindex_list_ij>::type,
+                       tmpl::integral_list<std::int32_t, 1, 1>>);
+
+  CHECK(std::is_same_v<typename TensorExpressions::detail::AddSubSymmetry<
+                           symmetric_symm, symmetric_symm, tensorindex_list_ij,
+                           tensorindex_list_ji>::type,
+                       tmpl::integral_list<std::int32_t, 1, 1>>);
+
+  CHECK(std::is_same_v<typename TensorExpressions::detail::AddSubSymmetry<
+                           asymmetric_symm, symmetric_symm, tensorindex_list_ij,
+                           tensorindex_list_ij>::type,
+                       tmpl::integral_list<std::int32_t, 2, 1>>);
+
+  CHECK(std::is_same_v<typename TensorExpressions::detail::AddSubSymmetry<
+                           asymmetric_symm, symmetric_symm, tensorindex_list_ij,
+                           tensorindex_list_ji>::type,
+                       tmpl::integral_list<std::int32_t, 2, 1>>);
+
+  CHECK(std::is_same_v<typename TensorExpressions::detail::AddSubSymmetry<
+                           symmetric_symm, asymmetric_symm, tensorindex_list_ij,
+                           tensorindex_list_ij>::type,
+                       tmpl::integral_list<std::int32_t, 2, 1>>);
+
+  CHECK(std::is_same_v<typename TensorExpressions::detail::AddSubSymmetry<
+                           symmetric_symm, asymmetric_symm, tensorindex_list_ij,
+                           tensorindex_list_ji>::type,
+                       tmpl::integral_list<std::int32_t, 2, 1>>);
+}
+
+void test_rank3() {
```

Comment:
what do you think of adding some `if (symmetry<...> != ...) { ERROR("It appears AddSub and the symmetry are longer using the same assumptions. You must update AddSub to assume the same symmetry things."); }` (with better words) so that the assumption stated in the comment is verified and we know it's been broken?

Reply:
Ah, so something like this for the input symmetry aliases?:
```
  using symm_111 = Symmetry<1, 1, 1>;
  static_assert(std::is_same_v<symm_121, tmpl::integral_list<std::int32_t, 1, 2, 1>, "(message)">);
  using symm_121 = Symmetry<1, 2, 1>;
  static_assert(std::is_same_v<symm_111, tmpl::integral_list<std::int32_t, 1, 1, 1>, "(message)">);
  using symm_211 = Symmetry<2, 1, 1>;
  ...
```
or for the the output symmetries?:
```
  static_assert(std::is_same_v<Symmetry<2, 2, 1>, tmpl::integral_list<std::int32_t, 2, 2, 1>>, "message");
  CHECK(std::is_same_v<typename TensorExpressions::detail::AddSubSymmetry<
                           symm_111, symm_121, tensorindex_list_abc,
                           tensorindex_list_bca>::type,
                       tmpl::integral_list<std::int32_t, 2, 2, 1>>);
  ...
```
or both?

In general, I think this is a really good idea

Reply:
Yea, exactly. Just one or two checks to verify the ordering assumption remains satisfied/synchronized between the two locations :)

Reply:
Will do!

Reply:
In the fixup, I put several general `ERROR` checks that are run up front and cover a few different cases, but don't exhaustively check all the test cases in the `CHECK`s below. I can alternatively do the latter, if you prefer.